test/Constraints/closures.swift

```diff
@@ -295,9 +295,9 @@ func simplified1069() {
     func genericallyNonOptional<T: AnyObject>(_ a: T, _ b: T, _ c: T) { }
 
     func f(_ a: C?, _ b: C?, _ c: C) {
-      genericallyNonOptional(a, b, c) // expected-error 2{{value of optional type 'C?' must be unwrapped to a value of type 'C'}}
-      // expected-note @-1 2{{coalesce}}
-      // expected-note @-2 2{{force-unwrap}}
+      genericallyNonOptional(a, b, c) // expected-error {{value of optional type 'C?' must be unwrapped to a value of type 'C'}}
```

Comment:
How many errors does it actually produce right now? It's supposed to be one for `a` and one for `b`.

Reply:
I can see that right now it produces 4 errors, twice for every argument, so maybe new changes are producing only one per?

Reply:
@xedin I think 2, but I am not sure as the `// expected-error <number>` syntax is new to me. If I change it to `// expected-error 1`, the test still passes. If I change it to `// expected-error 2`, then it complains.

Reply:
Could you just try this one snippet in repl or something and see what it produces?

Reply:
@xedin Without the `// expected-{error/note}`, it produces

> unexpected error produced: value of optional type 'C?' must be unwrapped to a value of type 'C'
> unexpected note produced: coalesce using '??' to provide a default when the optional value contains 'nil'
> unexpected note produced: force-unwrap using '!' to abort execution if the optional value contains 'nil'

With `// expected-error` and `// expected-error 1`, it passes. With `// expected-error 2` or more, it fails.

Reply:
Hm, which means that it only diagnoses once now for the first argument only. `expected-error 2` means that it expects two errors with the same text associated with a particular line. So original `expected-error 2` behavior  was correct since it was one error per invalid argument. I guess once we've removed the path, raw locator anchor for arguments now points to the call itself.

Here is an idea - there is `simplifyLocatorToAnchor` utility, let's try to use that when constructing `ForceOptional` fix, I'd suggest you modify `ForceOptional` constructor to do something like `ConstraintFix(root, cs, simplifyLocatorToAnchor(cs, locator)), ...` and revert new code from `ForceOptional::create` calls. 

`simplifyLocatorToAnchor` should return anchor for each argument if you give it locator like `[Call -> apply argument -> compare argument #0 to parameter #0]` which should fix this problem because `ForceOptional` is going to be associated with invalid argument expression instead of a top-level call expression.

Reply:
@xedin Oh! For some reason, I am unable to do that:

```c++
ConstraintFix(cs, FixKind::ForceOptional, simplifyLocatorToAnchor(cs, locator)), BaseType(baseType), UnwrappedType(unwrappedType) 
```

I can see `simplifyLocatorToAnchor` in code completion but I get `Use of undeclared identifier 'simplifyLocatorToAnchor'` when I use it in the constructor. 🤔 

Also, `ConstraintFix` expects `CS`, `FixKind` and `ConstraintLocator`, but simplifyLocatorToAnchor() returns `Expr`. I suppose I will have to extract the locator from that?

Reply:
You'd have to move constructor to `CSFix.cpp` because `CSFix.h` doesn't include `ConstraintSystem.h` on purpose and for the Expr* problem do cs.getConstraintLocator(simplifyLocatorToAnchor(cs, locator))` I forgot that it doesn't return full locator...

Reply:
@xedin Ah ok, seems like I can simply do `return new (cs.getAllocator()) ForceOptional(cs, baseType, unwrappedType, cs.getConstraintLocator(simplifyLocatorToAnchor(cs, locator)));` and it works.

It does undo some diagnostic messages that I recently changed. For example,

```
// Before this change
value of optional type 'Int?' must be unwrapped to a value of type 'Int'
// After this change - back to old message
[Int?]' to expected argument type 'Set<Int>
```

Reply:
Sure, that works as well! Just need to make sure that constructor is private.

Regarding the old messages now being the same - don't worry about it, these diagnostics are reasonable too.

Reply:
@xedin Okay, done :) Could you invoke the CI again? I think all the tests should pass now!